### DISCUSSION TITLE
Making OneOf feature not require strickNullCheck to be enabled.

### DISCUSSION THREAD

**Discussion by jimmymkude** [timostamm/protobuf-ts#254] [Ideas]
I played around with this, and I was eventually was able to get it work with ('strictNullChecks': false) by the changing the generated ts file to use a class to define the undefined oneof object instead of { oneofKind: undefined; }.

Using the docs example:

message OneofExample {
 // Only one of (or none of) the fields can be set
 oneof result { 
   int32 value = 1;
   string error = 2;
 }
}

interface OneofExample {
 result: { oneofKind: "value"; value: number; } 
       | { oneofKind: "error"; error: string; } 
       | { oneofKind: undefined; };
}

This would change to:

class UndefinedOneOf {
  oneofKind: undefined;
}

interface OneofExample {
 result: { oneofKind: "value"; value: number; } 
       | { oneofKind: "error"; error: string; } 
       | UndefinedOneOf;
}

Now if we handle the undefined case first, it gives us the type guard to narrow down the type. For example, in the generated internalByteWrite function.

if (message.result instanceof UndefinedOneOf)
  // Type guard above narrows down scope.
else if (message.result.oneofKind === "value")
  // Knows the result type is not undefined.
  Number.internalBinaryWrite(message.result.value, writer.tag(4, WireType.LengthDelimited).fork(), options).join();

What are your thoughts?

**Comment by timostamm** [comment]
Hey Jimmy, this is a valid workaround, but probably has some drawbacks around `switch (x.oneofKind)`?

It looks like `strickNullChecks` is enabled by default if you enable `strict`, which is greatly recommended. This was probably changed in one of the TypeScript versions in the past 1 or 2 years. I've updated the manual, thanks for raising this.

**Reply by jimmymkude** [reply]
Yes good point, I have verified that it doesn't work with switch statements since they don't provide the scope narrowing the if else provides.
